### PR TITLE
CB-17566: [pre-checks] Implement new flow for pre-checks needed for backup restore.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CB-17566

This is now tested and ready for merging once the reviews are in.

Summary of changes:

- Added a new DL service flow called `BackupValidation` which essentially just calls the Thunderhead DR API in order to run the validation flow present there, and waits for it with a generic poller
- Added a new DL service flow chain for DL Upgrade Preparation which includes the new backup validation flow and the original DL upgrade preparation flow
- Minor test changes to accommodate the changes 

Testing done:

- Added unit tests to verify that the flow chain is created correctly
- Manually set up CB locally and ran the prepare upgrade process using the exposed REST endpoints, ensuring that the backup validation was hit and worked correctly

### CODE REVIEW

**Review by kkalvagadda1 — APPROVED** [review]
Santiago, 

Please update the summary with the testing that was done on this change.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java [hunk 1]
```diff
@@ -187,12 +188,23 @@ public FlowIdentifier triggerDatabaseServerUpgradeFlow(SdxCluster cluster, Targe
...
-        LOGGER.info("Trigger Datalake runtime upgrade preparation for: {} with imageId: {}", cluster, imageId);
+    public FlowIdentifier triggerDatalakeRuntimeUpgradePreparationFlow(SdxCluster cluster, String imageId, boolean skipBackup) {
+        LOGGER.info("Trigger Datalake runtime upgrade preparation for: {} with imageId: {}", cluster.getClusterName(), imageId);
         String userId = ThreadBasedUserCrnProvider.getUserCrn();
-        return notify(DATALAKE_UPGRADE_PREPARATION_TRIGGER_EVENT.event(),
-                new DatalakeUpgradePreparationStartEvent(DATALAKE_UPGRADE_PREPARATION_TRIGGER_EVENT.event(), cluster.getId(),
-                        userId, imageId), cluster.getClusterName());
+        if (!skipBackup && sdxBackupRestoreService.shouldSdxBackupBePerformed(
+                cluster, entitlementService.isDatalakeBackupOnUpgradeEnabled(ThreadBasedUserCrnProvider.getAccountId())
+        )) {
```
Comment (lturcsanyi) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:196]: Code formatting

File: datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    @Bean(name = "DATALAKE_TRIGGERING_BACKUP_VALIDATION_STATE")
+    public Action<?, ?> triggerDatalakeBackupValidationAction() {
+        return new AbstractSdxAction<>(DatalakeTriggerBackupValidationEvent.class) {
+            @Override
+            protected void doExecute(SdxContext context, DatalakeTriggerBackupValidationEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Triggering datalake backup validation for {}", payload.getResourceId());
+                variables.put(REASON, payload.getReason().name());
+
+                SdxCluster sdxCluster = sdxService.getById(payload.getResourceId());
+                sdxStatusService.setStatusForDatalakeAndNotify(DatalakeStatusEnum.RUNNING,
```
Comment (lturcsanyi) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java:69]: Is it intentional to set RUNNING status?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                        ResourceEvent.DATALAKE_BACKUP_VALIDATION_IN_PROGRESS,
+                        "Datalake backup validation in progress", payload.getResourceId());
+                metricService.incrementMetricCounter(MetricType.SDX_BACKUP_VALIDATION_REQUESTED, sdxCluster);
+
+                DatalakeBackupStatusResponse backupStatusResponse =
```
Comment (lturcsanyi) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java:74]: If its a long running task, e.g: a http/grcp call, it should be moved to a Handler class.
See: https://github.com/hortonworks/cloudbreak/tree/master/flow#handlers

File: datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java (hunk 2 above)
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java:69]: Yes, the idea is that the validation, as well as other backup operations, do not affect the DL as a whole but simply serve as background operations that are occurring while the DL is still up and running.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java (hunk 3 above)
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/dr/validation/DatalakeBackupValidationActions.java:74]: Should it? There are plenty of other places that do this exact same thing. Also all this does is fire a GRPC call without waiting for it, instead we opt to have a waiting process as a separate step in the flow.
Hopefully that makes sense.

**Review by lturcsanyi — APPROVED** [review]

**Pull request merged**